trlx/trainer/nemo/gpt.py

```diff
@@ -0,0 +1,878 @@
+
+    stage_layers = num_layers // pp_size
+    pp_offset = pp_rank * stage_layers
+
+    encoder_layers_key = "model.language_model.encoder.layers."
+
+    def filter_in_pp_rank(key):
+        if key.startswith(encoder_layers_key):
+            layer_idx = int(key.split(".")[4])
+            return pp_offset <= layer_idx < (pp_offset + stage_layers)
+        elif key.startswith(
+            "model.language_model.encoder.final_layernorm"
+        ) and not pp_rank == (pp_size - 1):
+            return False
+        else:
+            return True
+
+    def shift_layer_idx(key):
+        """If the key is for a transformer layer, shift down the layer index to select the
+        correct layer for this pipeline stage."""
+        if key.startswith(encoder_layers_key):
+            layer_idx = int(key.split(".")[4])
+            return f"{encoder_layers_key}{str(layer_idx - pp_offset)}.{'.'.join(key.split('.')[5:])}"
+        else:
+            return key
+
+    state_dict = {
+        shift_layer_idx(k): v for k, v in state_dict.items() if filter_in_pp_rank(k)
+    }
+
+    return state_dict
+
+
+class ILQLGPT(MegatronGPTModel):
+    ilql_config: ILQLConfig
+
+    def __init__(self, ilql_config, metric_fn=None, **kwargs):
+        self.ilql_config = ilql_config
+        self.metric_fn = metric_fn
+        super().__init__(**kwargs)
+        if len(list(self.parameters())) == 0:
+            raise ValueError("No parameters in model")
+
+        self._ori_activations_checkpoint_granularity = self.cfg.get(
+            "activations_checkpoint_granularity", None
+        )
+        self._ori_activations_checkpoint_method = self.cfg.get(
+            "activations_checkpoint_method", None
+        )
+        self._ori_activations_checkpoint_num_layers = self.cfg.get(
+            "activations_checkpoint_num_layers", None
+        )
+
+    @classmethod
+    def list_available_models(cls) -> Optional[Mapping[str, str]]:
+        return None
+
+    def build_train_valid_test_datasets(self):
+        pass
+
+    def build_data_loader(self, dataset, collate_fn, consumed_samples=0):
+        dp_rank = parallel_state.get_data_parallel_rank()
+        dp_size = parallel_state.get_data_parallel_world_size()
+        print(
+            f"Building data loader for {type(dataset)=} {len(dataset)=} {dp_rank=} {dp_size=}",
+            file=sys.stderr,
+        )
+        batch_sampler = MegatronPretrainingBatchSampler(
+            total_samples=len(dataset),
+            consumed_samples=consumed_samples,
+            micro_batch_size=self.cfg.micro_batch_size,
+            global_batch_size=self.cfg.global_batch_size,
+            data_parallel_rank=dp_rank,
+            data_parallel_size=dp_size,
+            drop_last=True,
+        )
+        return torch.utils.data.DataLoader(
+            dataset,
+            batch_sampler=batch_sampler,
+            # For some reason this causes a crash when using >0 workers
```

Comment:
If `_reconfigure` for generate is uncommented, as currently is, `num_workers=self.cfg.data.num_workers` seems to work fine. Do you think we should change it back from `0`?